pkg/keyspace/tso_keyspace_group.go

```diff
@@ -106,8 +237,77 @@ func (m *GroupManager) saveKeyspaceGroups(keyspaceGroups []*endpoint.KeyspaceGro
 			if oldKG != nil {
 				return ErrKeyspaceGroupExists
 			}
+			newKG := &endpoint.KeyspaceGroup{
+				ID:       keyspaceGroup.ID,
+				UserKind: keyspaceGroup.UserKind,
+				Replica:  keyspaceGroup.Replica,
+			}
 			m.store.SaveKeyspaceGroup(txn, newKG)
 		}
 		return nil
 	})
 }
+
+func (m *GroupManager) updateKeyspaceGroup(keyspaceGroupID, keyspaceID uint32) error {
+	return m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
+		// Check if keyspace group has already existed.
+		oldKG, err := m.store.LoadKeyspaceGroup(txn, keyspaceGroupID)
+		if err != nil {
+			return err
+		}
+		if oldKG == nil {
+			return errKeyspaceGroupNotFound
+		}
+		newKG := &endpoint.KeyspaceGroup{
+			ID:        oldKG.ID,
+			UserKind:  oldKG.UserKind,
+			Replica:   oldKG.Replica,
+			Keyspaces: append(oldKG.Keyspaces, keyspaceID),
+		}
+		m.store.SaveKeyspaceGroup(txn, newKG)
+		return nil
+	})
+}
+
+// GetNodesNum returns the number of nodes.
+func (m *GroupManager) GetNodesNum() int {
+	return len(m.nodesBalancer.GetAll())
+}
+
+// AddKeySpace adds a keyspace to the keyspace group.
+func (m *GroupManager) AddKeySpace(keyspace uint32, userKind string) (uint32, error) {
+	// TODO: consider the case that the keyspace group offline
+	m.groups.RLock()
+	defer m.groups.RUnlock()
+	if balancer, ok := m.groups.balancerMap[userKind]; ok {
+		keyspaceGroup := balancer.Next()
+		if keyspaceGroup == 0 {
+			return 0, errKeyspaceGroupNotFound
+		}
+		if err := m.updateKeyspaceGroup(keyspaceGroup, keyspace); err != nil {
+			return 0, err
+		}
+		return keyspaceGroup, nil
+	}
+	return 0, errKeyspaceGroupKindNoExist
+}
+
+// Note: the replica should be less than the number of nodes.
+func (m *GroupManager) allocNodes(replica int) []endpoint.KeyspaceGroupMember {
+	exists := make(map[string]struct{})
+	nodes := make([]endpoint.KeyspaceGroupMember, replica)
+	for i := 0; i < replica; i++ {
+		addr := m.nodesBalancer.Next()
```

Comment:
Next advances the index, but it returns nil here if the node is not enough.

Reply:
`Next` will return "" when there is no node. So this function will return nil.

Reply:
Consider the case that we use 1,3,1,3,1,3... as the request and there are only two nodes, all requests with 1 replica will be allocated in the same node?

Reply:
the node of the balancer is deduplicated. And it will return "" when the balancer is empty and the internal index will not increase.